File: 405_to_hex/to_hex.cpp
/*
 * @Date: 2021-10-02 08:34:19
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-10-02 08:39:01
 */

#include <cassert>
#include <string>

using namespace std;

class Solution {
 public:
  string toHex(int num) {
    if (num == 0) return "0";

    string sb;
    for (int i = 7; i >= 0; i--) {
      int val = (num >> (4 * i)) & 0xf;
      if (sb.length() > 0 || val > 0) {
        char digit = val < 10 ? (char)('0' + val) : (char)('a' + val - 10);
        sb.push_back(digit);
      }
    }
    return sb;
  }
};

int main() {
  assert(Solution().toHex(26) == "1a");
  assert(Solution().toHex(-1) == "ffffffff");
}

File: golang/405_to_hex/to_hex.go
/*
 * @Date: 2021-10-02 08:34:22
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-10-02 08:46:50
 */

package main

import "strings"

func toHex(num int) string {
	if num == 0 {
		return "0"
	}
	sb := &strings.Builder{}
	for i := 7; i >= 0; i-- {
		val := num >> (4 * i) & 0xf
		if val > 0 || sb.Len() > 0 {
			var digit byte
			if val < 10 {
				digit = '0' + byte(val)
			} else {
				digit = 'a' + byte(val-10)
			}
			sb.WriteByte(digit)
		}
	}
	return sb.String()
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	assert(toHex(26) == "1a")
	assert(toHex(-1) == "ffffffff")
}
